src/hardware/shared_adc.rs

```diff
@@ -0,0 +1,113 @@
+///
+/// ## Synchronization
+/// If the multiple priorities utilize the ADC that results in resource pre-emption, pre-emption is
+/// protected against through the use of an atomic bool. Attempting to utilize the ADC from a
+/// higher priority level while it is in use at a lower level will result in a [AdcError::InUse].
+use embedded_hal::adc::{Channel, OneShot};
+use stm32h7xx_hal as hal;
+
+#[derive(Debug, Copy, Clone)]
+pub enum AdcError {
+    /// Indicates that the ADC channel has already been allocated.
+    Allocated,
+
+    /// Indicates that the ADC is already in use
+    InUse,
+}
+
+/// A single channel on an ADC peripheral.
+pub struct AdcChannel<'a, Adc, PIN> {
+    pin: PIN,
+    slope: f32,
+    mutex: &'a spinning::Mutex<hal::adc::Adc<Adc, hal::adc::Enabled>>,
+}
+
+impl<'a, Adc, PIN> AdcChannel<'a, Adc, PIN>
+where
+    PIN: Channel<Adc, ID = u8>,
+    hal::adc::Adc<Adc, hal::adc::Enabled>: OneShot<Adc, u32, PIN>,
+    <hal::adc::Adc<Adc, hal::adc::Enabled> as OneShot<Adc, u32, PIN>>::Error:
+        core::fmt::Debug,
+{
+    /// Read the ADC channel.
+    ///
+    /// # Returns
+    /// The normalized ADC measurement as a ratio of full-scale.
+    pub fn read(&mut self) -> Result<f32, AdcError> {
+        let mut adc = self.mutex.try_lock().ok_or(AdcError::InUse)?;
+        Ok(adc.read(&mut self.pin).unwrap() as f32 / self.slope)
+    }
+}
+
+/// An ADC peripheral that can provide ownership of individual channels for sharing between
+/// drivers.
+pub struct SharedAdc<Adc> {
+    mutex: spinning::Mutex<hal::adc::Adc<Adc, hal::adc::Enabled>>,
+    allocated_channels: core::cell::RefCell<[bool; 20]>,
+    slope: f32,
+}
+
+impl<Adc> SharedAdc<Adc> {
+    /// Construct a new shared ADC driver.
+    ///
+    /// # Args
+    /// * `slope` - The slope of the ADC conversion transfer function.
+    /// * `adc` - The ADC peripheral to share.
+    pub fn new(slope: f32, adc: hal::adc::Adc<Adc, hal::adc::Enabled>) -> Self {
+        Self {
+            slope,
+            mutex: spinning::Mutex::new(adc),
+            allocated_channels: core::cell::RefCell::new([false; 20]),
+        }
+    }
+
+    /// Allocate an ADC channel for usage.
+    ///
+    /// # Args
+    /// * `pin` - The ADC input associated with the desired ADC channel. Often, this is a GPIO pin.
+    ///
+    /// # Returns
+    /// An instantiated [AdcChannel] whose ownership can be transferred to other drivers.
+    pub fn create_channel<PIN: Channel<Adc, ID = u8>>(
+        &self,
+        pin: PIN,
+    ) -> Result<AdcChannel<'_, Adc, PIN>, AdcError> {
+        let mut channels = self.allocated_channels.borrow_mut();
+        if channels[PIN::channel() as usize] {
+            return Err(AdcError::Allocated);
+        }
+
+        channels[PIN::channel() as usize] = true;
+
+        Ok(AdcChannel {
+            pin,
+            slope: self.slope,
+            mutex: &self.mutex,
+        })
+    }
+}
+
+macro_rules! new_shared_adc {
```

Comment:
And can't that be just inlined? Doesn't look like a lot of code.

Reply:
Yeah, there's a fair argument for making the code more understandable. Let me take a look.